src/sink/kafka.rs

```diff
@@ -0,0 +1,214 @@
+    /// Total record publish failures.
+    pub static ref KAFKA_PUBLISH_FAILURE_SUM: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
+}
+
+/// Config options for Kafka config.
+#[derive(Clone, Debug, Deserialize)]
+pub struct KafkaConfig {
+    /// Canonical name for the given Kafka sink.
+    pub config_path: Option<String>,
+    /// Kafka topic to publish to.
+    pub topic_name: Option<String>,
+    /// Kafka brokers. This is a comma-separated list of host or host:port.
+    pub brokers: Option<String>,
+    /// Underlying librdkafka configuration.
+    pub rdkafka_config: Option<HashMap<String, String>>,
+    /// How often (seconds) the in-flight messages are checked for delivery.
+    /// Default = 1 second
+    pub flush_interval: u64,
+}
+
+impl Default for KafkaConfig {
+    fn default() -> KafkaConfig {
+        KafkaConfig {
+            config_path: None,
+            topic_name: None,
+            brokers: None,
+            rdkafka_config: None,
+            flush_interval: 1,
+        }
+    }
+}
+
+/// Kafka sink internal state.
+pub struct Kafka {
+    /// Name of the stream we are publishing to.
+    topic_name: String,
+    /// A message producers.
+    producer: FutureProducer<EmptyContext>,
+    // In-flight messages.
+    messages: Vec<DeliveryFuture>,
+    /// How often (seconds) the in-flight messages are checked for delivery.
+    flush_interval: u64,
+}
+
+impl Sink<KafkaConfig> for Kafka {
+    fn init(config: KafkaConfig) -> Self {
+        if config.topic_name.is_none() {
+            panic!("No Kafka topic name provided!");
+        }
+        if config.brokers.is_none() {
+            panic!("No Kafka brokers provided!")
+        }
+
+        let mut producer_config = ClientConfig::new();
+        if let Some(ref map) = config.rdkafka_config {
+            for (key, value) in map.iter() {
+                producer_config.set(key, value);
+            }
+        }
+        producer_config.set("bootstrap.servers", &config.brokers.unwrap()[..]);
+
+        Kafka {
+            topic_name: config.topic_name.unwrap(),
+            producer: producer_config.create::<FutureProducer<_>>().unwrap(),
+            messages: Vec::new(),
+            flush_interval: config.flush_interval,
+        }
+    }
+
+    fn valve_state(&self) -> Valve {
+        // We never close up shop.
+        Valve::Open
+    }
+
+    fn deliver(&mut self, _: Arc<Option<Telemetry>>) -> () {
+        // Discard point
+    }
+
+    fn deliver_line(&mut self, _: Arc<Option<LogLine>>) -> () {
+        // Discard line
+    }
+
+    /// Fire off the given event to librdkafka. That library handles buffering and
+    /// batching internally.
+    fn deliver_raw(
+        &mut self,
+        order_by: u64,
+        _encoding: metric::Encoding,
+        bytes: Vec<u8>,
+    ) {
+        let key = format!("{:X}", order_by);
+        let future = self.try_payload(bytes.as_slice(), key.as_bytes());
+        self.messages.push(future);
```

Comment:
Why don't we bound on the size of messages here?

What saves us from blowing up a system by sending N events in a flush interval where N is 1 more event than the system can handle?

Reply:
There is now configurable valve control through `max_message_bytes` -- once the valve closes, deliver_raw stops getting called. So, we're now bounding (just not based on message count).